### PR TITLE
Output event logs for failed live tests

### PR DESCRIPTION
**Pull request by JoshLove-msft**
Peeling off the event logging changes from https://github.com/Azure/azure-sdk-for-net/pull/9150 into separate PR, so we can merge this in while working on resolution for DataLake errors.

### CODE REVIEW

File: sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.cs [hunk 1]
```diff
@@ -28,13 +29,69 @@ static StorageTestBase()
...
+        }
+
+        /// <summary>
+        /// Stop logging events and do necessary cleanup.
+        /// This will run once after all tests have finished.
+        /// </summary>
+        [OneTimeTearDown]
+        public void StopLoggingEvents()
+        {
+            s_listener?.Dispose();
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.cs]: Nit - null it out after disposing so your other checks for null won't ever make calls on a disposed instance.

File: sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.cs [hunk 2]
```diff
@@ -28,13 +29,69 @@ static StorageTestBase()
...
+        /// <summary>
+        /// Sets up the Event listener buffer for the test about to run.
+        /// This will run prior to the start of each test.
+        /// </summary>
+        [SetUp]
+        public void SetupEventsForTest()
+        {
+            if (s_listener != null)
+            {
+                TestEventListener.SetupEventsForTest();
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.cs]: Why isn't SetupEventsForTest an instance method?

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs [hunk 3]
```diff
@@ -21,21 +22,38 @@ namespace Azure.Storage.Test
     /// </summary>
     internal class TestEventListener : AzureEventSourceListener
     {
+        private static StringBuilder s_eventBuffer;
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs]: It feels weird to create instances but dump their output in a static buffer.  Why not make this an instance member?

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs [hunk 4]
```diff
@@ -79,6 +97,10 @@ public static void LogEvent(EventWrittenEventArgs args)
 
             // Dump the message and category
             Trace.WriteLine(message, category);
+
+            // Add the message to event buffer
+            s_eventBuffer.Append(message);
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs]: Do we check if this is null before using it?

**Review by tg-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.cs (hunk 1 above)
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.cs]: Fixed.

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs]: The reason for all of the statics is that we need to pass an Action to the base ctor in TestEventListener ctor. This forces the method used in the action to be a static.

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs (hunk 4 above)
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs]: It shouldn't be null if that method is executing. If it is null, that would indicate a bug somewhere so I'd rather not add a check.

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs (hunk 3 above)
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs]: Should we instead be deriving from the TestEventListener class in Core.Testing? That class doesn't seal OnEventWritten so we could override it to do our logging.

Comment (pakrym) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs]: Don't inherit from  AzureEventSourceListener, instantiate it in your ctor.

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs (hunk 4 above)
Comment (tg-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs]: Then add an Assert it's not null with a message that you need to call the setup method first.  Somebody's going to mess this up in the future and it'd be nice to help them out when they do.

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs [hunk 5]
```diff
@@ -19,23 +20,45 @@ namespace Azure.Storage.Test
...
+        /// Output the Events to the console in the case of test failure.
+        /// This will include the HTTP requests and responses.
+        /// </summary>
+        public void OutputEventsForTest()
         {
-            if (!Debugger.IsAttached)
+            if (TestContext.CurrentContext.Result.Outcome.Status == TestStatus.Failed)
             {
-                return;
+                TestContext.Progress.WriteLine(_eventBuffer.ToString());
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs]: Why TestContext.Progress instead of TestContext.Error?

**Review by tg-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs (hunk 5 above)
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs]: I think Error should work. Progress is a holdover from when I was outputting events for all tests. I will need to test it out though to make sure it works as expected.

**Comment by JoshLove-msft** [comment]
@heaths live storage tests seem to be failing on "Create test resources" step.
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=218062&view=logs&j=96d7b680-dcef-5cfd-cab1-6c247ea8b6c3&t=2c71fa7c-1f8f-5833-9aff-8c6966b5622b

Should this be skipped until Storage is updated to use ARM templates?

**Review by seanmcc-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs (hunk 5 above)
Comment (JoshLove-msft) [reply sdk/storage/Azure.Storage.Common/tests/Shared/TestEventListener.cs]: Error ends up printing the logs twice. It seems to get printed immediately and then again as part of the error information.

File: sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.cs [hunk 6]
```diff
@@ -28,13 +29,60 @@ static StorageTestBase()
 #endif
         }
 
+        /// <summary>
+        /// Add a static TestEventListener which will redirect SDK logging
+        /// to Console.Out for easy debugging.
```
Comment (tg-msft) [inline sdk/storage/Azure.Storage.Common/tests/Shared/StorageTestBase.cs]: Nit - Update the comment to talk about dumping failures as this no longer dumps all logging.

**Pull request merged**